src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy

```diff
@@ -0,0 +1,888 @@
+                        def resources = traffExam.startExam(direction)
+                        direction.setResources(resources)
+                        assert traffExam.waitExam(direction).hasTraffic()
+                    }
+        }
+
+        when: "Update the QinQ flow"
+        def updateResponse = flowHelperV2.updateFlow(flow.flowId, flow.tap {
+            flow.source.vlanId = flow2.source.vlanId
+            flow.source.innerVlanId = flow2.destination.vlanId
+            flow.destination.vlanId = flow2.destination.vlanId
+            flow.destination.innerVlanId = flow2.source.vlanId
+        })
+
+        then: "Update response contains correct info about innerVlanIds"
+        with(updateResponse) {
+            it.source.vlanId == flow2.source.vlanId
+            it.source.innerVlanId == flow2.destination.vlanId
+            it.destination.vlanId == flow2.destination.vlanId
+            it.destination.innerVlanId == flow2.source.vlanId
+        }
+
+        and: "Flow is really updated"
+        with(northbound.getFlow(flow.flowId)) {
+            it.source.vlanId == flow2.source.vlanId
+            it.source.innerVlanId == flow2.destination.vlanId
+            it.destination.vlanId == flow2.destination.vlanId
+            it.destination.innerVlanId == flow2.source.vlanId
+        }
+
+        then: "Both existing flows are still valid and pingable"
+        [flow.flowId, flow2.flowId].each {
+            northbound.validateFlow(it).each { assert it.asExpected }
+        }
+
+        [flow.flowId, flow2.flowId].each {
+            verifyAll(northbound.pingFlow(it, new PingInput())) {
+                it.forward.pingSuccess
+                it.reverse.pingSuccess
+            }
+        }
+
+        when: "Delete the flows"
+        [flow.flowId, flow2.flowId].each { flowHelperV2.deleteFlow(it) }
+
+        then: "Flows rules are deleted"
+        involvedSwitchesforBothFlows.each { sw ->
+            Wrappers.wait(RULES_INSTALLATION_TIME, 1) {
+                assert northbound.getSwitchRules(sw.dpId).flowEntries*.cookie.sort() == sw.defaultCookies.sort()
+            }
+        }
+
+        and: "Shared rule of flow is deleted"
+        [swP.src.dpId, swP.dst.dpId].each { swId ->
+            assert northbound.getSwitchRules(swId).flowEntries.findAll {
+                new Cookie(it.cookie).getType() == CookieType.SHARED_OF_FLOW
+            }.empty
+        }
+
+        where:
+        srcVlanId | srcInnerVlanId | dstVlanId | dstInnerVlanId
+        0         | 0              | 0         | 0
+        10        | 20             | 30        | 40
+        10        | 10             | 10        | 10
+        10        | 0              | 0         | 40
+        0         | 20             | 30        | 0
+        10        | 20             | 0         | 0
+        0         | 0              | 30        | 40
+        10        | 20             | 30        | 0
+        0         | 20             | 30        | 40
```

Comment:
Nah. too many, we can't afford this for such a long-running scenario. While pairwising you should consider that src and dst for us are basically equal entities. I think it can be reduced to
```
        srcVlanId | srcInnerVlanId | dstVlanId | dstInnerVlanId
        0         | 1              | 1         | 0   
       1          | 1              | 2        | 1
```
1row: [noVlan,Vlan] tested on src, [vlan,noVlan] tested on dst, tested transition between src/dst outer: noVlan->Vlan, inner: vlan ->noVlan
2row: [vlan,sameVlan] tested on src, [vlan,otherVlan] tested on dst, tested transition between src/dst outer: vlan->sameVlan, inner: vlan ->otherVlan

We can further debate on 0-0-0-0 and 1-1-1-1
Update operation does not really update anything for them and actual cases are already covered, but I don't mind having one/both of these additional 2 iterations for the sake of traffic check, maybe as low_pro...or not have them, your call

Reply:
Let's do whatever set we need while the feature is in development, but I don't think we can afford dragging so many permutations to daily jobs. After we stabilize with testing qnq we should review this set or at least mark less valuable variations as low_prio. Otherwise, this only test will consume most of our execution time. Also, we can later consider moving most permutations to a lighter test which is not doing 2 updates + 3 traffexam sessions per each iteration